MetaDataset.py

```diff
@@ -550,75 +552,147 @@ def __init__(self,
 
-    for dataset in self.datasets.values():
-      dataset.init_seq_order(epoch=epoch)
     return True
 
+  def _create_dataset_seq_idx_list(self):
+    """
+    Creates a list of all available sequences, to which we can later apply the sequence ordering.
+    It contains the index of the dataset and the sequence index within the dataset for every sequence.
+    The sequences appear sorted by dataset first.
+
+    :returns: list of (dataset-idx, dataset-seq-idx)
+    :rtype: list[(int,int)]
+    """
+    dataset_seq_idxs = []
+    for dataset_idx in range(len(self.datasets)):
+      dataset = self.datasets[self.dataset_idx2key_map[dataset_idx]]
+      for dataset_seq_idx in range(dataset.num_seqs):
+        dataset_seq_idxs.append((dataset_idx, dataset_seq_idx))
+
+    return dataset_seq_idxs
+
+  def _get_random_dataset_seq_order(self):
+    """
+    Choose datasets randomly but preserve order within each dataset. This sorting method is unique to CombinedDataset.
+    """
+    # Create a list containing each dataset_idx dataset.num_seqs-times and shuffle it.
+    dataset_ids = [idx_tuple[0] for idx_tuple in self.dataset_seq_idx_list]
+    self.rnd.shuffle(dataset_ids)
+
+    # Calculate the offset for all datasets (i.e. the index where sequences of a given dataset start
+    # in self.dataset_seq_idx_list).
+    dataset_offsets = [0]
+    for dataset_idx in range(len(self.datasets)):
+      dataset = self.datasets[self.dataset_idx2key_map[dataset_idx]]
+      dataset_offsets.append(dataset_offsets[-1] + dataset.num_seqs)
+
+    # Create the actual seq_order list.
+    # We want to keep the order within the sub-datasets, thus we assign seq_ids by simply counting up for each dataset.
+    # We however have to account for the different offsets needed when accessing self.dataset_seq_idx_list later.
+    seq_order = []
+    counters = [0] * len(self.datasets)
+    for dataset_idx in dataset_ids:
+      seq_order.append(counters[dataset_idx] + dataset_offsets[dataset_idx])
+      counters[dataset_idx] += 1
+
+    total_num_seqs = len(self.dataset_seq_idx_list)
+    assert sum(counters) == total_num_seqs
+
+    if self.partition_epoch:
+      seq_order = self._apply_partition_epoch(seq_order, self.partition_epoch, total_num_seqs, self.epoch)
+    if self.repeat_epoch:
+      seq_order = seq_order * self.repeat_epoch
+
+    return seq_order
+
+  def _get_seq_length(self, seq_idx):
+    dataset_idx, dataset_seq_idx = self.dataset_seq_idx_list[seq_idx]
+    dataset = self.datasets[self.dataset_idx2key_map[dataset_idx]]
+
+    return dataset.get_seq_length(dataset_seq_idx)["data"]
+
   def _expand_dataset_sec_idxs(self, num_values):
     """
     :param num_values: int Add num_values entries to the dataset-segment-idx mapping table
     :return:
     """
     for i in range(num_values):
-      if self.seq_ordering in ("default", "random"):  # default is random. this is different from base class!
+      if self.seq_ordering == "default":
```

Comment:
"default" and "random" are now different things. "random" shuffles all sequences, while "default" keeps the order of the sub-datasets and just randomly picks the dataset.

Reply:
Maybe we should give it another explicit name, like "random_dataset".

Reply:
Yes. "default" should behave as it behaves in all other datasets. It would be confusing otherwise, I think.
This case, "random_dataset", do you really need that? Do you think this is useful? I'm not sure. Don't add this if you don't need it now.

Reply:
Ok, I just tried to keep the old behaviour. So then I make 'in-order' the default, i.e. to go through all sequences of dataset1, then dataset2 and so on.
I didn't add 'random_dataset', this is the only case that was there before ;) And it is useful, for example if one dataset has much longer sequence lengths than the other (i.e. text vs. audio). Then a joint 'laplace' sorting for training makes no sense, but you can still sort each dataset and then pick datasets randomly.